### ISSUE TITLE
CA1416 warnings are erroneously shown in lambdas which are only reachable after platform checks

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn-analyzers#7323]
### Analyzer

**Diagnostic ID**: [CA1416](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1416): Validate platform compatibility

### Analyzer source

**Version**: SDK 8.0.300

This is particularly agonizing for APIs such as the host/services builders.

The below should not show warnings, but it does:

```cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

if (!OperatingSystem.IsWindows()) return 1;

await Host.CreateDefaultBuilder(args)
    .ConfigureServices((hostContext, services) =>
    {
        services.AddLogging(builder =>
        {
            // ⚠️ CA1416: This call site is reachable on all platforms. 'EventLoggerFactoryExtensions.AddEventLog(
            // ILoggingBuilder, Action<EventLogSettings>)' is only supported on: 'windows'.
            builder.AddEventLog(eventLogSettings =>
            {
                // ⚠️ CA1416: This call site is reachable on all platforms. 'EventLogSettings.SourceName' is only
                // supported on: 'windows'.
                eventLogSettings.SourceName = "Foo";
            });
        });
    }).Build().RunAsync();
return 0;
```

The workaround makes it clear that this is an inconsistency in the analyzer's behavior; an implicit "SupportedOSPlatform" should be applied to all statements following `if (!OperatingSystem.IsWindows()) return 1;` without manually stating this:

```cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using System.Runtime.Versioning;

if (!OperatingSystem.IsWindows()) return 1;

[SupportedOSPlatform("windows")] void Configure(HostBuilderContext hostContext, IServiceCollection services)
{
    services.AddLogging(builder =>
    {
        builder.AddEventLog(eventLogSettings =>
        {
            eventLogSettings.SourceName = "Foo";
        });
    });
}

await Host.CreateDefaultBuilder(args).ConfigureServices(Configure).Build().RunAsync();
return 0;
```

**Comment by buyaa-n** [comment]
Closing as duplicate of https://github.com/dotnet/roslyn-analyzers/issues/4282, which is closed as by design 

The flow analysis could not evaluate a lambda expression unless it is explicitly invoked. Please check the [issue]( https://github.com/dotnet/roslyn-analyzers/issues/4282) for details

> The workaround makes it clear that this is an inconsistency in the analyzer's behavior; an implicit "SupportedOSPlatform" should be applied to all statements following if (!OperatingSystem.IsWindows()) return 1; without manually stating this:

The suggested workaround is to add a Debug.Assert at the start of the lambda for the expected OS information.

**Comment by jnm2** [comment]
@buyaa-n Thanks. I asked for clarification [in the linked issue](https://github.com/dotnet/roslyn-analyzers/issues/4282#issuecomment-2334131162).
